pip_audit/dependency_source/resolvelib/pypi_provider.py

```diff
@@ -110,6 +121,34 @@ def get_metadata_for_wheel(url):
     return EmailMessage()  # pragma: no cover
 
 
+def get_metadata_for_sdist(url):
+    data = requests.get(url).content
+
+    # Extract archive onto the disk
+    with TarFile.open(fileobj=BytesIO(data), mode="r:gz") as t:
+        # The directory is the first member in a tarball
+        names = t.getnames()
+        pkg_dir = names[0]
+        t.extractall()
+
+    cwd = os.getcwd()
```

Comment:
I'm planning to add some code to generate an unused directory name and extract the tarball and put the virtualenv in there just so we don't get clashes if the user just so happens to have something called `tmp_env` in their cwd.

But this is the general gist of what I'm intending to do. I'm making a fresh virtualenv, installing the sdist on it and then doing a `pip freeze` to figure out what packages got installed. One issue (which I think is pretty minor) is that this code is reporting back some seed packages like `setuptools`, `pip`, etc.

Reply:
This general approach looks reasonable to me!

Instead of building the virtualenv in the current directory, we should probably use the system temporary directory by way of the `tempfile` module. In particular, we can do some naive key-style caching by using a scheme like this:

```
${TMPDIR}/pip-audit-${some-hash-of-requirements-input}/
```

...and if the hash hasn't changed, we can assume that the state under that directory is up-to-date and not recompute it. That in turn involves leaving temporary directories around, however, so maybe it isn't worth it.

Reply:
Using temporary directories in 66e6c7cb170483273c1bf96c828b49009cb3c31a.

I decided to just keep it simple for now and have it use temporary directories for the sdist extraction and the virtual environment. I figure that if performance turns out to be a problem for real life `requirements.txt`, then we can introduce caching at a later point.